modules/imgcodecs/src/grfmt_tiff.cpp

```diff
@@ -565,90 +530,136 @@ bool  TiffEncoder::writeLibTiff( const Mat& img, const std::vector<int>& params)
+    for (std::vector<Mat>::const_iterator it = img_vec.begin(); it != img_vec.end(); ++it)
     {
-        TIFFClose(pTiffHandle);
-        return false;
-    }
+        int channels = it->channels();
+        int width = it->cols, height = it->rows;
+        int depth = it->depth();
 
-    for (int y = 0; y < height; ++y)
-    {
-        switch(channels)
+        int bitsPerChannel = -1;
+        switch (depth)
         {
-            case 1:
+            case CV_8U:
             {
-                memcpy(buffer, img.ptr(y), scanlineSize);
+                bitsPerChannel = 8;
                 break;
             }
-
-            case 3:
+            case CV_16U:
             {
-                if (depth == CV_8U)
-                    icvCvt_BGR2RGB_8u_C3R( img.ptr(y), 0, buffer, 0, cvSize(width,1) );
-                else
-                    icvCvt_BGR2RGB_16u_C3R( img.ptr<ushort>(y), 0, (ushort*)buffer, 0, cvSize(width,1) );
-                break;
-            }
-
-            case 4:
-            {
-                if (depth == CV_8U)
-                    icvCvt_BGRA2RGBA_8u_C4R( img.ptr(y), 0, buffer, 0, cvSize(width,1) );
-                else
-                    icvCvt_BGRA2RGBA_16u_C4R( img.ptr<ushort>(y), 0, (ushort*)buffer, 0, cvSize(width,1) );
+                bitsPerChannel = 16;
                 break;
             }
-
             default:
             {
-                TIFFClose(pTiffHandle);
                 return false;
             }
         }
 
-        int writeResult = TIFFWriteScanline(pTiffHandle, buffer, y, 0);
-        if (writeResult != 1)
+
+        const int bitsPerByte = 8;
+        size_t fileStep = (width * channels * bitsPerChannel) / bitsPerByte;
+
+        int rowsPerStrip = (int)((1 << 13) / fileStep);
+        readParam(params, TIFFTAG_ROWSPERSTRIP, rowsPerStrip);
+
+        if (rowsPerStrip < 1)
+            rowsPerStrip = 1;
+
+        if (rowsPerStrip > height)
+            rowsPerStrip = height;
+
+        int   colorspace = channels > 1 ? PHOTOMETRIC_RGB : PHOTOMETRIC_MINISBLACK;
+
+        if (!TIFFSetField(pTiffHandle, TIFFTAG_IMAGEWIDTH, width)
+            || !TIFFSetField(pTiffHandle, TIFFTAG_IMAGELENGTH, height)
+            || !TIFFSetField(pTiffHandle, TIFFTAG_BITSPERSAMPLE, bitsPerChannel)
+            || !TIFFSetField(pTiffHandle, TIFFTAG_COMPRESSION, compression)
+            || !TIFFSetField(pTiffHandle, TIFFTAG_PHOTOMETRIC, colorspace)
+            || !TIFFSetField(pTiffHandle, TIFFTAG_SAMPLESPERPIXEL, channels)
+            || !TIFFSetField(pTiffHandle, TIFFTAG_PLANARCONFIG, PLANARCONFIG_CONTIG)
+            || !TIFFSetField(pTiffHandle, TIFFTAG_ROWSPERSTRIP, rowsPerStrip)
+            )
         {
             TIFFClose(pTiffHandle);
             return false;
         }
+
+        if (compression != COMPRESSION_NONE && !TIFFSetField(pTiffHandle, TIFFTAG_PREDICTOR, predictor))
+        {
+            TIFFClose(pTiffHandle);
+            return false;
+        }
+
```

Comment:
Do we also need to set fields `TIFFTAG_SUBFILETYPE` (to `FILETYPE_PAGE`) and `TIFFTAG_PAGENUMBER` (with args `currentPageNumber`, `numberOfPages`)?

Reply:
It seems to work fine without these fields. But, I believe these fields will be helpful nonetheless.